### PR TITLE
Add off-diagonal jacobian of phase-field fracture variable respect to displacements

### PR DESCRIPTION
**Pull request by jiangwen84**
closes ref #11790

### CODE REVIEW

**Comment by jiangwen84** [comment]
@dschwen any suggestion to make this kernel more general? thanks!

**Comment by jiangwen84** [comment]
@dschwen any comments? thanks!!!

File: modules/phase_field/include/kernels/AllenCahnElasticEnergyOffDiag.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+class AllenCahnElasticEnergyOffDiag
+  : public DerivativeMaterialInterface<JvarMapKernelInterface<Kernel>>
+{
+public:
+  AllenCahnElasticEnergyOffDiag(const InputParameters & parameters);
+
+protected:
+  Real computeQpResidual() override { return 0.0; }
+
+  Real computeQpJacobian() override { return 0.0; }
```
Comment (dschwen) [inline modules/phase_field/include/kernels/AllenCahnElasticEnergyOffDiag.h]: No need to override that one, it defaults to `return 0.0;` (see `KernelBase`)

File: modules/phase_field/include/kernels/AllenCahnElasticEnergyOffDiag.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
+
+  /// Mobility
+  const MaterialProperty<Real> & _L;
```
Comment (dschwen) [inline modules/phase_field/include/kernels/AllenCahnElasticEnergyOffDiag.h]: Just a remark: technically symbol names consisting of a leading underscore followed by a single uppercase letter are reserved for internal use of the compiler. We've encountered cases where this breaks the build of MOOSE (on Windows).

We have plenty of those in MOOSE but we really shouldn't add more.

File: modules/phase_field/src/kernels/AllenCahnElasticEnergyOffDiag.C [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+registerMooseObject("PhaseFieldApp", AllenCahnElasticEnergyOffDiag);
+
+template <>
+InputParameters
+validParams<AllenCahnElasticEnergyOffDiag>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("This kernel calculates off-diagonal jacobian of elastic energy in "
+                             "AllenCahn respect to displacements");
```
Comment (dschwen) [inline modules/phase_field/src/kernels/AllenCahnElasticEnergyOffDiag.C]: `with respect to`

File: modules/phase_field/src/kernels/AllenCahnElasticEnergyOffDiag.C [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  params.addCoupledVar("displacements",
+                       "The string of displacements suitable for the problem statement");
```
Comment (dschwen) [inline modules/phase_field/src/kernels/AllenCahnElasticEnergyOffDiag.C]: `vector of displacements` maybe?

File: modules/phase_field/src/kernels/AllenCahnElasticEnergyOffDiag.C [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  params.addParam<MaterialPropertyName>(
+      "F_name", "E_el", "Name of material property storing the elastic energy");
+  params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
```
Comment (dschwen) [inline modules/phase_field/src/kernels/AllenCahnElasticEnergyOffDiag.C]: I'd like to move away from naming material property parameters `*_name`. The user can supply a plain number here, not just a name.

**Comment by jiangwen84** [comment]
After discussion with @dschwen , the comments regarding the material properties' name will be addressed in a separate PR. Other comments are addressed.

**Comment by friedmud** [comment]
@dschwen can you take a final look at this?

**Review by dschwen — APPROVED** [review]

**Pull request merged**